tests/test_restriction.cpp

```diff
@@ -0,0 +1,288 @@
+ *  We hence only consider 6 adjacent fine nodes, which lie on one of the edges spanned by the corresponding coarse node. 
+ *  This is matrix-free but it corresponds to applying the transposed prolongation matrix
+ *
+ *  Parametrized tests are used to test for different grid sizes and with or without Dirichlet boundary conditions.
+ */
+
+TEST_P(test_restriction, test_extrapolation_restriction)
+{
+    gyro::icntl[Param::nr_exp]         = (int)(std::get<0>(GetParam()) / 3) + 3;
+    gyro::icntl[Param::ntheta_exp]     = (std::get<0>(GetParam()) % 3) + 3;
+    gyro::icntl[Param::DirBC_Interior] = std::get<1>(GetParam());
+    if (gyro::icntl[Param::nr_exp] == 3)
+        gyro::icntl[Param::fac_ani] = 2; //anisotropy should not exceed grid size
+
+    gmgpolar test_p;
+
+    create_grid(test_p);
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int;
+    int cr_int     = test_p.v_level[1]->nr_int;
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta;
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta;
+
+    std::vector<double> u_test(p_level.m);
+    for (int z = 0; z < p_level.m; z++) {
+        u_test[z] = 1 - z; //arbitrary grid function
+    }
+
+    std::vector<double> sol = p_level.apply_restriction_ex(u_test);
+
+    /* based on the triangulation, at most 6 adjacent fine nodes are considered:
+     * **left,top_left,bottom,top,bottom_right,right**
+     */
+    for (int j = 0; j < cr_int + 1; j++) {
+        for (int i = 0; i < ctheta_int; i++) {
+            std::vector<double> adjacent(6, -1.0);
+            if (j == 0) { //interior circle
+                adjacent[0] = 0.0;
+                adjacent[1] = 0.0;
+            }
+            if (j == cr_int) { //exterior circle
+                adjacent[4] = 0.0;
+                adjacent[5] = 0.0;
+            }
+
+            int k = 0;
+            //relative values of the fine nodes as in bilinear restriction
+            for (int z = -1; z < 2; z++) {
+                for (int y = -1; y < 2; y++) {
+                    if ((z != 0 || y != 0) && (z * y != 1)) {
+                        if (adjacent[k] != 0.0) {
+
+                            adjacent[k] = u_test[(2 * j + z) * p_level.ntheta_int +
+                                                 ((i != 0 || y > -1) ? (2 * i + y) : p_level.ntheta_int - 1)];
+                        }
+                        k += 1;
+                    }
+                }
+            }
+
+            double finval = u_test[(2 * j) * p_level.ntheta_int + (2 * i)];
+            for (int z = 0; z < 8; z++) {
+                finval +=
+                    0.5 *
+                    adjacent[z]; //accumulate the values. the vector "vals" reduces to 1/2 for every adjacent fine node
+            }
+
+            EXPECT_NEAR(finval, sol[j * ctheta_int + i], 1e-6)
```

Comment:
```suggestion
            EXPECT_NEAR(finval, sol[j * ctheta_int + i], 1e-12)
```
Can we make this succeed with 1e-12 too?